---
title: "Phyloseq_Preprocessing"
author: "Rena Fukuda"
output: html_document
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "figures/02_Phyloseq_Preprocessing")
```

# Goals of this file

Process data into phyloseq object. 
Removing the following: 

1. Mitochondria 
2. Chloroplasts 
3. Samples without enough reads
4. ASVs from negative controls

Then, evaluate accuracy by looking at mock community samples. 
Finally, write out data file of phyloseq output, which will be saved in
"data/Phyloseq/raw_preprocessed_physeq.RData"

# Load libraries

```{r load_libraries}
#BiocManager::install("phyloseq")
pacman::p_load(devtools, phyloseq, tidyverse, dada2, install = FALSE)
```

# Load data
## Load ASV Table
```{r load_asv_table}
# Load in data
load("data/01_DADA2/ASV_counts.RData")

# Look at first five cols
head(asv_tab[,1:5])

# Fix names
sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1) 
head(samples_fixed)

# Rewrite ASV count file to fix names
colnames(asv_tab) <- samples_fixed

# Check structure
str(asv_tab)
```

## Load Taxonomy Table
```{r load_tax_table}
tax_df <- read.table("data/01_DADA2/ASV_taxonomy.tsv", sep = "\t", header = TRUE)
head(tax_df)

tax_df_new <- tax_df[,-1]
rownames(tax_df_new) <- tax_df[,1]
tax_df_new <- as.matrix(tax_df_new)
```

## Load Track Read Counts
```{r load_track_counts}
# Load data
load("data/01_DADA2/track_read_counts.RData")

# Look at data
head(track_counts_df)
dim(track_counts_df)
```

```{r load_metadata}
# Load in metadata
metadata_df <- read.csv("data/02_Phyloseq/metadata.csv")

# Look at data
dim(metadata_df)
colnames(metadata_df)

# Merge metadata with track reads
metadata_track_df <- metadata_df %>% 
  left_join(., track_counts_df, by = "names") %>% 
  dplyr::select(-X)

# Intuition check 
head(metadata_track_df)

# Update row names to be sample names
# Before
row.names(metadata_track_df)
# Rewrite
row.names(metadata_track_df) <- metadata_track_df$names
# Check
row.names(metadata_track_df)

# Intuition check
head(metadata_track_df)

```

# Handoff to phyloseq
```{r handoff_phyloseq}
# Double check
dim(asv_tab)
dim(tax_df_new)

# Intuition check
stopifnot(row.names(asv_tab) == row.names(tax_df_new))

# Construct phyloseq object
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE),
                       sample_data(metadata_track_df),
                       tax_table(tax_df_new)) 

# Save raw physeq
save(raw_physeq, file = "data/02_Phyloseq/raw_physeq.RData")
```

# Clean up data
Remove:
1. Chloroplasts
2. Mitochondria

```{r remove_mito_chloro}
# Make new physeq w/o mito (family) or chloro (order)
# Rm chloroplast
no_chloro_physeq <- 
  raw_physeq %>%
    subset_taxa((Order != "Chloroplast") | is.na(Order))

# How many were chloro?
num_chloro_asvs <- ntaxa(raw_physeq) - ntaxa(no_chloro_physeq)
num_chloro_asvs

# Intuition check
# no_chloro_physeq %>% 
#   tax_table %>%
#   data.frame() %>%
#   view()

# Remove mitochondria
no_chloro_mito_physeq <- 
  no_chloro_physeq %>%
    subset_taxa((Family != "Mitochondria") | is.na(Family))

# How many mitochondrial asvs?
num_mito_asvs <- ntaxa(no_chloro_physeq) - ntaxa(no_chloro_mito_physeq)
num_mito_asvs

# How many were removed?
ntaxa(raw_physeq) - ntaxa(no_chloro_mito_physeq)
# Proportion of ASVs left?
ntaxa(no_chloro_mito_physeq)/ntaxa(raw_physeq)
```

# Evaluate and remove control samples

Take a look at negative controls and decide whether or not to remove ASVs found in controls.

1. Negative controls.
2. ASVs found within negative controls and their distribution in samples.
3. Evaluate mock community

## Negative controls
```{r neg_controls}
# Create a vector of samples that were neg controls
control_samples <- c("WaterControl", "022um-Control",
                    "3um-Control", "DNA-Ext-Control")

# Make new phyloseq of controls
control_physeq <- 
  no_chloro_mito_physeq %>% 
  subset_samples(., names %in% control_samples) %>%
  # force remove ASVs with zero counts
  prune_taxa(taxa_sums(.) >0, .)

control_physeq

# Make vector of control ASVs
control_asvs <- 
  control_physeq %>%
  tax_table() %>%
  data.frame %>%
  dplyr::select(ASV) %>%
  as.vector()

# Visualize raw abundance values of control ASVs
control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")

# Check control asvs in samples
# Make new physeq obj with only control asvs and all samples
control_asv_physeq <- 
  no_chloro_mito_physeq %>%
  subset_taxa(., ASV %in% control_asvs$ASV)

control_asv_physeq

# Take a look at abundances
control_asv_physeq %>%
  plot_bar(., "ASV", fill = "ASV", facet_grid = Sample_or_Control~.)

# Check tax of ASV3
# control_asv_physeq %>% 
#   tax_table() %>%
#   data.frame() %>%
#   view()

# Remove eveything but ASV18 and ASV56
control_asvs_to_remove <- dplyr::setdiff(control_asvs$ASV, c("ASV_18", "ASV_56"))
# check
length(control_asvs_to_remove)

# Make new phyloseq object without control asvs or control samples
no_mito_chloro_ctrl_physeq <- 
  no_chloro_mito_physeq %>%
  subset_taxa(., !(ASV %in% control_asvs_to_remove)) %>%
  subset_samples(., !(names %in% control_samples))

# Inspect no control object
no_mito_chloro_ctrl_physeq

# How many ASVs are removed?
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(no_mito_chloro_ctrl_physeq)
total_ASVs_rm

```

# Evaluate accuracy in the Mock community
- make a mock phyloseq object
- load in the standard fasta sequences of the mock community
- test for differences between the two

```{r mock_community}
# Make mock phyloseq object
mock_physeq <- no_mito_chloro_ctrl_physeq %>%
  subset_samples(., names == "MockZymoPos") %>%
  prune_taxa(taxa_sums(.) > 0 , .)

# Inspect it
mock_physeq
tax_table(mock_physeq)

# Write a message summarizing 
cat("DADA2 inferred", ntaxa(mock_physeq), "ASVs present in the mock community.")

# Load in mock refernce sequence from Zymo
mock_ref <- getSequences("data/02_Phyloseq/mock_amplicons.fasta")
names(mock_ref)

# Pull out seqs from mock community
mock_seqs <- mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseqs)

# Inspect
mock_seqs

# Which ASVs match the reference? (match is 0/1 if match)
match_ASVs <- mock_seqs %>% 
  rowwise() %>%
  mutate(Match = any(grep(ASVseqs, mock_ref))) %>%
  pull(ASV)

cat(length(match_ASVs), "ASVs were exact matches to the expected reference sequence")

# How many seqs were in mock?
mock_seq_depth <- mock_physeq %>% 
  otu_table() %>%
  data.frame() %>%
  colSums()

mock_seq_depth

# Who is in seq mock comm?
mock_physeq %>%
  tax_table() %>%
  data.frame() %>%
  dplyr::select(Genus, Species)
names(mock_ref)
```

What can we conclude about the mock comm?
```{r physeq_wo_mock}
no_ctrls_mock_mito_chloro <- no_mito_chloro_ctrl_physeq %>% 
  subset_samples(., names != "MockZymoPos") %>%
  subset_taxa(., !(ASV %in% match_ASVs))

no_ctrls_mock_mito_chloro

# Intuition check 
num_ASVs_mock = ntaxa(no_mito_chloro_ctrl_physeq) - ntaxa(no_ctrls_mock_mito_chloro)
num_ASVs_mock

stopifnot(num_ASVs_mock == 8)
```

# Evalulate te sequencing depth
```{r seq_depth}
# Curr data obj
no_ctrls_mock_mito_chloro

# Library depth per sample
seq_sums_df <- no_ctrls_mock_mito_chloro %>%
  otu_table() %>%
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>% 
  left_join(., metadata_track_df, by = "names") 

# Check
dim(seq_sums_df)
head(seq_sums_df)

# Rename second col
colnames(seq_sums_df)[2] <- "Totalseqs"

# Show depth of samples
seq_sums_df %>%
  dplyr::select(names, Totalseqs) %>%
  arrange(Totalseqs) %>%
  head

# Plot it
seq_sums_df %>%
  ggplot(aes(x = reorder(names, Totalseqs), y=Totalseqs, fill = station)) +
  geom_bar(stat = "Identity") 

# Density plot
seq_sums_df %>%
  ggplot(aes(Totalseqs, fill = station)) +
  geom_density(alpha = 0.5) 
  
```

# Remove samples with few reads
```{r remove_low_read_samp}
# Min seq depth
min(sample_sums(no_ctrls_mock_mito_chloro))

# Remove 20210615-MA-SCB2F 
row_preprocessed_physeq <- no_ctrls_mock_mito_chloro %>%
  subset_samples(., names != "20210615-MA-SCB2F" )

# Double check min seq depth
min(sample_sums(row_preprocessed_physeq))
```

# Save preprocessed phyloseq obj
```{r save_physeq}
save(row_preprocessed_physeq, file = "data/02_Phyloseq/raw_preprocessed_physeq.RData")

```


# Session Information 
```{r session-info}
# Ensure reproducibility 
devtools::session_info()
```



