---
title: "05_Between_Sample_Diversity"
author: "Rena Fukuda"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.path = "../figures/05_Community_Analysis/")
```

# Goals

1. Load in phyloseq data with rooted true.
2. Evaluate sequencing depth and remove a sample.
3. Normalize the read counts between samples. 
4. Calculate community disimilarity indices 
  a. *Sorenson:* Shared species as a binary value (abundance-unweighted)
  b. *Bray-Curtis:* Shared abundant species (abundance-weighted)
  c. *weighted UniFrac:* Consider abundant species and branch lenghts 
5. Visualize the commmunity data with two unconstrained ordinations.
  a. *PCoA:* Linear method, eigenvalue: how much variation is explained by each axis
  b. *NMDS:* Multiple dimensions into 2 or 3
6. Run statistics with PERMANOVA and betadispR.

# Set up environment
## Load libraries
```{r load_libraries}
#install.packages("vegan")
pacman::p_load(tidyverse, devtools, vegan, phyloseq, patchwork, install = FALSE)
```

## Set seed
```{r set_seed}
set.seed(123456)
```

## Load in data
```{r load_physeq}
# Load in rooted phylogenetic tree
load("data/03_Phylogenetic_Tree/Phytree_Preprocessed_Physeq.RData")
```

```{r load_colors}
station_colors <- c(
  "Shipping Channel" = "dodgerblue4",
  "Aransas Bay" = "dodgerblue2",
  "Copano West" = "#D9CC3C",
  "Copano East" = "#A0E0BA",
  "Mesquite Bay" = "#00ADA7")
```

# Explore read counts
```{r calc_seq_depth}
# Calculate total number of reads per sample
raw_total_seqs_df <- midroot_physeq_rm456 %>%
  # Calculate the sample read sums
  sample_sums() %>%
  data.frame() 

colnames(raw_total_seqs_df)[1] <- "TotalSeqs"

# Make a histogram of raw reads
ggplot(raw_total_seqs_df, aes(x = TotalSeqs)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(limits = c(0,10000)) +
  labs(title = "Raw Sequencing Depth Distribution") +
  theme_bw()
```


## Normalize read counts

# Calculate & visualize community dissimilarity
## Sorenson PCoA

## Bray-Curtis PCoA

## Weighted UniFrac PCoA

## Bray-Curtis NMDS

# Test for statistical significance with PERMANOVA and betadispR
# Session info
```{r session_info}
devtools::session_info()
```

